libs/core/src/lib/input-group/input-group-directives.ts

```diff
@@ -5,30 +5,42 @@ import {
     HostBinding,
     Input,
     OnChanges,
+    OnDestroy,
     OnInit,
+    Optional,
     Renderer2
 } from '@angular/core';
 import { InputGroupPlacement } from './input-group.component';
 import { FormStates } from '../form/form-control/form-states';
-import { applyCssClass, CssClassBuilder } from '../utils/public_api';
+import { applyCssClass, ContentDensityService, CssClassBuilder } from '../utils/public_api';
+import { Subscription } from 'rxjs';
 
 @Directive({
     // tslint:disable-next-line:directive-selector
     selector: '[fd-input-group-input]'
 })
-export class InputGroupInputDirective implements CssClassBuilder, OnInit, OnChanges {
+export class InputGroupInputDirective implements CssClassBuilder, OnInit, OnChanges, OnDestroy {
     /** user's custom classes */
     @Input()
     class: string;
 
     @Input()
-    compact = false;
+    compact: boolean = null;
 
     /** @hidden */
-    constructor(private _elementRef: ElementRef) {}
+    private _subscriptions = new Subscription();
+
+    /** @hidden */
+    constructor(private _elementRef: ElementRef, @Optional() private _contentDensityService: ContentDensityService) {}
 
     /** @hidden */
     ngOnInit(): void {
+        if (this.compact === null && this._contentDensityService) {
+            this._subscriptions.add(this._contentDensityService.contentDensity.subscribe(density => {
```

Comment:
taking in account change proposed to the service, `if` check can be removed and this line will be

```suggestion
            this._subscriptions.add(this?._contentDensityService?.subscribe(density => {
```

Reply:
what is more important here, is that you are loosing a default value for `compact`

if you move subscription to constructor, then value which is set from service will be overridden in `onInit`,
and then you can restore default value for compact

Reply:
I think you may have this backwards, what's happening right now with this PR is that the value set from the service is overridden whenever an input value is provided for the density (typically the [compact] input but occasionally [cozy]).  Is this not the desired behavior?  Moving this logic to the constructor then tells the component to listen to changes from the content density service and ignore whatever is there for the input.